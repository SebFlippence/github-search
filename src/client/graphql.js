import { GraphQLClient } from 'graphql-request';

const endpoint = 'https://api.github.com/graphql';

const client = new GraphQLClient(endpoint, {
  headers: {
    Authorization: `bearer ${process.env.REACT_APP_GITHUB_TOKEN}`
  }
});

export default client;
