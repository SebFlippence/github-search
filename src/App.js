import React, { PureComponent } from 'react';

import graphql from './client/graphql';
import SearchForm from './components/search/form';
import SearchResults from './components/search/results';
import urlUtils from './utils/url';
import userSearchQuery from './queries/github-user-search';

class App extends PureComponent {
  constructor(props) {
    super(props);
    this.state = { username: urlUtils.getUsername() };
  }

  async componentDidMount() {
    const { username } = this.state;
    if (username) {
      const variables = { username };

      this.setState({
        searchResults: await graphql.request(userSearchQuery, variables)
      });
    }
  }

  render() {
    const { username, searchResults } = this.state;
    return (
      <>
        <SearchForm search={username || 'e.g. JohnDoe'} />
        {username && (
          <>
            <hr />
            {searchResults ? (
              <SearchResults results={searchResults?.search?.nodes} />
            ) : (
              <p className="px-12 py-8 italic">Loading...</p>
            )}
          </>
        )}
      </>
    );
  }
}

export default App;
