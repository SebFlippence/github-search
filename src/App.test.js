import React from 'react';
import { render, wait } from '@testing-library/react';

import mockSearchResults from '../__mocks__/user-search-response';

import App from './App';
import urlUtils from './utils/url';
import graphql from './client/graphql';
import userSearchQuery from './queries/github-user-search';

jest.mock('./utils/url');
jest.mock('./client/graphql');

test('renders the search form', () => {
  const { baseElement } = render(<App />);
  expect(baseElement).toMatchSnapshot();
});

test('displays loading when a username is given, but data is not loaded', () => {
  urlUtils.getUsername.mockReturnValue('JohnDoe');
  const { baseElement, getByText } = render(<App />);
  const loadingElement = getByText('Loading...');
  expect(loadingElement).toBeInTheDocument();

  expect(baseElement).toMatchSnapshot();
});

test('displays SearchResults component when username and data is returned', async () => {
  urlUtils.getUsername.mockReturnValue('gaearon');
  graphql.request.mockResolvedValue(mockSearchResults);

  const { getByText, asFragment } = render(<App />);

  await wait(() => expect(getByText('Search results (3)')).toBeInTheDocument());

  expect(graphql.request).toHaveBeenCalledWith(userSearchQuery, {
    username: 'gaearon'
  });

  expect(asFragment()).toMatchSnapshot();
});
