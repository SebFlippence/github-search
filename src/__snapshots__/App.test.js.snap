// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays SearchResults component when username and data is returned 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="w-full max-w-md"
    >
      <form
        action=""
        class="px-8 py-8 pt-8"
      >
        <div
          class="px-4 pb-4"
        >
          <label
            class="text-xl block font-bold pb-2"
            for="username"
          >
            Search for a GitHub username
          </label>
          <input
            class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline border-blue-300 "
            id="username"
            name="username"
            placeholder="gaearon"
            type="username"
          />
        </div>
        <div
          class="px-4 pb-4"
        >
          <input
            class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
            type="submit"
            value="Search"
          />
        </div>
      </form>
    </div>
  </div>
  <hr />
  <div
    class="w-full max-w-md px-12 py-8"
  >
    <h2
      class="text-xl block font-bold pb-2 py-4"
    >
      Search results (3)
    </h2>
    <ul>
      <li
        class="py-3"
      >
        <button
          class="text-left flex items-center m-2"
          type="button"
        >
          <img
            alt="Avatar of gaearon"
            class="w-10 h-10 rounded-full mr-4"
            src="https://avatars0.githubusercontent.com/u/810438?v=4"
          />
          <div>
            <p
              class="text-gray-900 leading-none underline"
            >
              gaearon
            </p>
            <p
              class="text-gray-600"
            >
              Dan Abramov 
              <span
                class="italic"
              >
                <a
                  class="hover:underline"
                  href="mailto:dan.abramov@me.com"
                >
                  dan.abramov@me.com
                </a>
              </span>
            </p>
          </div>
        </button>
      </li>
      <li
        class="py-3"
      >
        <button
          class="text-left flex items-center m-2"
          type="button"
        >
          <img
            alt="Avatar of yashguptaz"
            class="w-10 h-10 rounded-full mr-4"
            src="https://avatars1.githubusercontent.com/u/36479718?v=4"
          />
          <div>
            <p
              class="text-gray-900 leading-none underline"
            >
              yashguptaz
            </p>
            <p
              class="text-gray-600"
            >
              Yash 
              <span
                class="italic"
              >
                <a
                  class="hover:underline"
                  href="mailto:yashguptaofficial@gmail.com"
                >
                  yashguptaofficial@gmail.com
                </a>
              </span>
            </p>
          </div>
        </button>
      </li>
      <li
        class="py-3"
      >
        <button
          class="text-left flex items-center m-2"
          type="button"
        >
          <img
            alt="Avatar of gaearoney"
            class="w-10 h-10 rounded-full mr-4"
            src="https://avatars3.githubusercontent.com/u/55377986?v=4"
          />
          <div>
            <p
              class="text-gray-900 leading-none underline"
            >
              gaearoney
            </p>
            <p
              class="text-gray-600"
            >
               
              <span
                class="italic"
              >
                <a
                  class="hover:underline"
                  href="mailto:"
                />
              </span>
            </p>
          </div>
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`displays loading when a username is given, but data is not loaded 1`] = `
<body>
  <div>
    <div
      class="App"
    >
      <div
        class="w-full max-w-md"
      >
        <form
          action=""
          class="px-8 py-8 pt-8"
        >
          <div
            class="px-4 pb-4"
          >
            <label
              class="text-xl block font-bold pb-2"
              for="username"
            >
              Search for a GitHub username
            </label>
            <input
              class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline border-blue-300 "
              id="username"
              name="username"
              placeholder="JohnDoe"
              type="username"
            />
          </div>
          <div
            class="px-4 pb-4"
          >
            <input
              class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
              type="submit"
              value="Search"
            />
          </div>
        </form>
      </div>
    </div>
    <hr />
    <p
      class="px-12 py-8 italic"
    >
      Loading...
    </p>
  </div>
</body>
`;

exports[`renders the search form 1`] = `
<body>
  <div>
    <div
      class="App"
    >
      <div
        class="w-full max-w-md"
      >
        <form
          action=""
          class="px-8 py-8 pt-8"
        >
          <div
            class="px-4 pb-4"
          >
            <label
              class="text-xl block font-bold pb-2"
              for="username"
            >
              Search for a GitHub username
            </label>
            <input
              class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline border-blue-300 "
              id="username"
              name="username"
              placeholder="e.g. JohnDoe"
              type="username"
            />
          </div>
          <div
            class="px-4 pb-4"
          >
            <input
              class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
              type="submit"
              value="Search"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</body>
`;
